portal-kernel/src/com/liferay/portal/kernel/util/CompaniesUtil.java

```diff
@@ -0,0 +1,186 @@
+		UnsafeFunction<Long, T, E> unsafeFunction,
+		BiConsumer<Long, E> biConsumer, Stream<Long> companyIdsStream) {
+
+		return companyIdsStream.flatMap(
+			companyId -> {
+				try (SafeClosable safeClosable =
+						CompanyThreadLocal.setWithSafeClosable(companyId)) {
+
+					CompanyThreadLocal.setCompanyId(companyId);
+
+					return Stream.of(unsafeFunction.apply(companyId));
+				}
+				catch (Exception exception) {
+					biConsumer.accept(companyId, (E)exception);
+
+					return Stream.empty();
+				}
+			});
+	}
+
+	public static <E extends Exception> void forEach(
+			UnsafeConsumer<Company, E> unsafeConsumer)
+		throws E {
+
+		forEach(unsafeConsumer, CompanyLocalServiceUtil.getCompanies(false));
+	}
+
+	public static <E extends Exception> void forEach(
+		UnsafeConsumer<Company, E> unsafeConsumer,
+		BiConsumer<Company, E> biConsumer) {
+
+		forEach(
+			unsafeConsumer, biConsumer,
+			CompanyLocalServiceUtil.getCompanies(false));
+	}
+
+	public static <E extends Exception> void forEach(
+		UnsafeConsumer<Company, E> unsafeConsumer,
+		BiConsumer<Company, E> biConsumer, List<Company> companies) {
+
+		for (Company company : companies) {
+			try (SafeClosable safeClosable =
+					CompanyThreadLocal.setWithSafeClosable(
+						company.getCompanyId())) {
+
+				unsafeConsumer.accept(company);
+			}
+			catch (Exception exception) {
+				biConsumer.accept(company, (E)exception);
+			}
+		}
+	}
+
+	@SuppressWarnings("all")
+	public static <E extends Exception> void forEach(
+			UnsafeConsumer<Company, E> unsafeConsumer, List<Company> companies)
+		throws E {
+
+		forEach(
+			unsafeConsumer, (__, e) -> ReflectionUtil.throwException(e),
+			companies);
+	}
+
+	public static <E extends Exception> void forEachCompanyId(
+			UnsafeConsumer<Long, E> unsafeConsumer)
+		throws E {
+
+		forEachCompanyId(unsafeConsumer, _getCompanyIds());
+	}
+
+	public static <E extends Exception> void forEachCompanyId(
+		UnsafeConsumer<Long, E> unsafeConsumer,
+		BiConsumer<Long, E> biConsumer) {
+
+		forEachCompanyId(unsafeConsumer, biConsumer, _getCompanyIds());
+	}
+
+	public static <E extends Exception> void forEachCompanyId(
+		UnsafeConsumer<Long, E> unsafeConsumer, BiConsumer<Long, E> biConsumer,
+		long[] companyIds) {
+
+		for (long companyId : companyIds) {
+			try (SafeClosable safeClosable =
+					CompanyThreadLocal.setWithSafeClosable(companyId)) {
+
+				unsafeConsumer.accept(companyId);
+			}
+			catch (Exception exception) {
+				biConsumer.accept(companyId, (E)exception);
+			}
+		}
+	}
+
+	@SuppressWarnings("all")
+	public static <E extends Exception> void forEachCompanyId(
+			UnsafeConsumer<Long, E> unsafeConsumer, long[] companyIds)
+		throws E {
+
+		forEachCompanyId(
+			unsafeConsumer, (__, e) -> ReflectionUtil.throwException(e),
+			companyIds);
+	}
+
+	private static long[] _getCompanyIds() {
+		long[] companyIds = new long[0];
+
+		try {
+			companyIds = PortalUtil.getCompanyIds();
+		}
+		catch (NullPointerException nullPointerException) {
```

Comment:
@achaparro this is too weird, catching for NPE?
Actually this entire class is weird, why is this not part of CompanyLocalService?

Reply:
@shuyangzhou, I catch the nullpointerException because the PortalUtil couldn't be initialized at that point. We can catch any Exception if it's better.

Regarding moving this to the CompanyLocalService, I haven't considered that because this logic is not part of the persistence layer, it is more a static class, we just need the companyIds. The only problem is that it is used in one or two bundles really early on startup, that's why we need to call the service.

Let me know what you think, we can have a call if you prefer it.

Thanks.

Reply:
The problem that it needs to use an uninitialized PortalUtil is a big sign showing it should not be using PortalUtil at all. This logic should be moved to CompanyLocalService since you are falling back to it anyway, I assume that will be ready by the time it is called. Service logic does not always tie to persistence layer, and in this case at least when doing the fallback it actually needs persistence layer. It is a right thing to do to move it CompanyLocalService.

Reply:
ok @shuyangzhou,

I will move the logic there. Do you want me to keep the usage of the PortalUtil to improve the performance or it's ok to call always companyLocalService.getCompanies because it will be cached?

Apart from that, do you see other things in the pull that I should review?

Reply:
No, just move everything to CompanyLocalService. If there is a performance issue, we can deal with it later. For now keep it clear first. I have not really started review yet, just spotted this for now. I do notice there are unused methods in the CompaniesUtil, please remove them when you move it to CompanyLocalService, only keep what is really used.e

Reply:
ok, let's do that. I will close the pull to prepare the changes.